pkg/controllers/metrics/node/controller.go

```diff
@@ -0,0 +1,338 @@
+/*
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package node
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"strings"
+
+	"knative.dev/pkg/logging"
+
+	"github.com/aws/karpenter/pkg/apis/provisioning/v1alpha5"
+	"github.com/prometheus/client_golang/prometheus"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/resource"
+	"k8s.io/apimachinery/pkg/types"
+	corev1client "k8s.io/client-go/kubernetes/typed/core/v1"
+	controllerruntime "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	crmetrics "sigs.k8s.io/controller-runtime/pkg/metrics"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+const (
+	nodeName         = "name"
+	nodeProvisioner  = "provisioner"
+	nodeZone         = "zone"
+	nodeArchitecture = "arch"
+	nodeCapacityType = "capacitytype"
+	nodeInstanceType = "instancetype"
+	nodePhase        = "phase"
+	nodeLabels       = "nodeLabels"
+)
+
+func getLabelNames() []string {
+	return []string{
+		nodeName,
+		nodeProvisioner,
+		nodeZone,
+		nodeArchitecture,
+		nodeCapacityType,
+		nodeInstanceType,
+		nodePhase,
+		nodeLabels,
+	}
+}
+
+type Controller struct {
+	KubeClient   client.Client
+	CoreV1Client corev1client.CoreV1Interface
+	LabelsMap    map[types.NamespacedName]*prometheus.Labels
+	GaugeVecMap  map[string]*prometheus.GaugeVec
```

Comment:
I typically see prometheus metrics stored as static variables, rather than member variables. It makes it easier to access metrics without a reference to an encapsulating context.

Reply:
done